src/init.cpp

```diff
@@ -965,17 +957,6 @@ void InitParameterInteraction()
             LogPrintf("%s: parameter interaction: -zapwallettxes=<mode> -> setting -rescan=1\n", __func__);
     }
 
-    if (GetBoolArg("-reindex", false) && !GetArg("-prune", 0)) {
-        if (SoftSetBoolArg("-rescan", true))
-            LogPrintf("%s: parameter interaction: -reindex=1 and not pruning -> setting -rescan=1\n", __func__);
-    }
```

Comment:
We do want to retain something that this soft-set was doing as a side-effect: clearing the wallet's witness cache (because otherwise as the block index and chain state are rebuilt, and the chain data is replayed to the wallet, the witnesses will get out-of-sync. We should do this by passing `clearWitnessCache || fReindex` into `InitLoadWallet` below.

The soft-set here was however misleading (in that it implies that `-reindex -rescan=0` could be useful, when in fact it was a config that would break the wallet more by not clearing the witness cache), and should indeed be reverted.

Reply:
I tried doing this in a pairing with @therealyingtong, and then realised that `CWallet::ClearNoteWitnessCache` is currently insufficient: the Sprout and Sapling witness caches should probably be cleared (though outside of a live reorg edge case I don't think it affects much), but the Orchard tree is not being cleared, and that is I think a separate bug in the current implementation. I'm going to do some local testing, and probably open a separate issue for this (so it won't block this PR).